server.js

```diff
@@ -9,25 +9,112 @@ import cors from 'cors'
 // import avocadoSalesData from './data/avocado-sales.json'
 // import booksData from './data/books.json'
 // import netflixData from './data/netflix-titles.json'
-// import topMusicData from './data/top-music.json'
+
+import topMusicData from './data/top-music.json';
 
 // Defines the port the app will run on. Defaults to 8080, but can be 
 // overridden when starting the server. For example:
 //
-//   PORT=9000 npm start
+//PORT=9000 npm start
 const port = process.env.PORT || 8080
 const app = express()
 
 // Add middlewares to enable cors and json body parsing
-app.use(cors())
-app.use(bodyParser.json())
+app.use(cors());
+app.use(bodyParser.json());
 
 // Start defining your routes here
-app.get('/', (req, res) => {
-  res.send('Hello world')
-})
+app.get('/', (request, response) => {
+  response.send('Welcome, /topmusic will show you the top tracks!')
+});
+
+// The first endpoint returns a collection of all top music 
+app.get('/topmusic', (request, response) => {
+  response.json(topMusicData);
+});
+
+// All artists
+app.get('/topmusic/artists', (request, response) => {
+  const artists = topMusicData.map((item) => item.artistName)
+  const uniqueArtists = [...new Set(artists)];
+
+  response.json(uniqueArtists);
+});
+
+// All tracks
+app.get('/topmusic/tracks', (request, response) => {
+  const tracks = topMusicData.map(item => item.trackName);
+  const uniqueTracks = [...new Set(tracks)];
+
+  response.json(uniqueTracks);
+});
+
+// Search by id 
+app.get('/topmusic/:id', (request, response) => {
+  const id = request.params.id
+  const topmusicID = topMusicData.find((item) => item.id === +id);
+  response.json(topmusicID);
+});
+
+// Search by trackname
+app.get('/topmusic/tracks/:trackName', (request, response, next) => {
+  const trackName = request.params.trackName;
+  const songs = topMusicData.filter((item) => item.trackName === trackName);
+  const tracks = songs.map(item => item.trackName);
+  const uniqueTracks = [...new Set(tracks)];
+
+  // If track is not found
+  if (uniqueTracks.length === 0) {
+    const error = new Error(`${trackName} not found`);
+    error.status = 404;
+    throw error;
+  };
+
+  response.json(uniqueTracks);
+});
```

Comment:
This doesn't have to be its own endpoint, wouldn't it be nice if this was a query param under your first endpoint? Something like:
`/songs?trackName="china"`

Super nice with error handling! You could improve the search function by making it case insensitive as well as letting the user search for a part of a track's name :)

Reply:
Also - wouldn't it be nice if the whole song object were returned ;)